server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,181 @@
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            if (event.localNodeMaster() && publishedAfterElection == false) {
+                submitHealthMetadata("health-metadata-update-master-election");
+            }
+            // If the node is not the elected master anymore
+            publishedAfterElection = event.localNodeMaster();
+        }
+    }
+
+    private void updateHealthMetadataIfNecessary() {
+        ClusterState clusterState = clusterService.state();
+        if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)
+            && clusterState.nodes().isLocalNodeElectedMaster()) {
+            submitHealthMetadata("health-metadata-update-threshold-change");
+        }
+    }
+
+    public static List<NamedXContentRegistry.Entry> getNamedXContentParsers() {
+        return List.of(
+            new NamedXContentRegistry.Entry(Metadata.Custom.class, new ParseField(HealthMetadata.TYPE), HealthMetadata::fromXContent)
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    private HealthMetadata createHealthMetadata() {
+        return new HealthMetadata(
+            HealthMetadata.DiskThresholds.createDiskThresholds(allocationDiskThresholdSettings, healthDiskThresholdSettings)
+        );
+    }
+
+    private void submitHealthMetadata(String source) {
+        submitHealthMetadata(source, 0);
+    }
+
+    // Visible for testing
+    void submitHealthMetadata(String source, int attempt) {
+        if (attempt < MAX_RETRIES) {
+            HealthMetadata localHealthMedata = createHealthMetadata();
+            String message = attempt == 0 ? source : "retry[" + attempt + "]-" + source;
+            var task = new UpdateHealthMetadataTask(localHealthMedata, () -> submitHealthMetadata(source, attempt + 1));
+            var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+            clusterService.submitStateUpdateTask(message, task, config, taskExecutor);
+        } else {
+            logger.error("Failed to process {} after {} attempts", source, attempt);
+        }
+    }
+
+    record UpdateHealthMetadataTask(HealthMetadata healthMetadata, Runnable retry) implements ClusterStateTaskListener {
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "never called";
+        }
+
+        @Override
+        public void onFailure(@Nullable Exception e) {
+            logger.error("unexpected failure during health metadata update", e);
+            retry.run();
```

Comment:
You are right, I checked on other cases and we simply log a failure. I was wondering if there could be cases where this fails and the metadata get out of sync for a long time since they do not get updated often. But if this scenario is rare, I would prefer too to remove the retry mechanism because it makes the code much more complex.